qcengine/programs/entos.py

```diff
@@ -0,0 +1,238 @@
+            return result
+
+    def execute(self, inputs, extra_outfiles=None, extra_commands=None, scratch_name=None, timeout=None):
+
+        if extra_outfiles is not None:
+            outfiles = ["dispatch.out"].extend(extra_outfiles)
+        else:
+            outfiles = ["dispatch.out"]
+
+        if extra_commands is not None:
+            commands = extra_commands
+        else:
+            commands = inputs["commands"]
+
+        exe_success, proc = execute(commands,
+                                    infiles=inputs["infiles"],
+                                    outfiles=outfiles,
+                                    scratch_location=inputs["scratch_location"],
+                                    scratch_name=scratch_name,
+                                    timeout=timeout
+                                    )
+        proc["outfiles"]["dispatch.out"] = proc["stdout"]
+
+        # Determine whether the calculation succeeded
+        output_data = {}
+        if not exe_success:
+            output_data["success"] = False
+            output_data["error"] = {"error_type": "internal_error",
+                                    "error_message": proc["stderr"]
+                                    }
+            return FailedOperation(
+                success=output_data.pop("success", False), error=output_data.pop("error"), input_data=output_data)
+
+        return proc
+
+    def build_input(self, input_model: 'ResultInput', config: 'JobConfig',
+                    template: Optional[str] = None) -> Dict[str, Any]:
+
+        # Ensure (optional dependency) jinja2 exists if template is provided
+        if template is not None and not which_import('jinja2', return_bool=True):
+            raise ModuleNotFoundError("""Python module jinja2 not found. Solve by installing it: `conda install 
+            jinja2` or `pip install jinja2`""")
+
+        # Write the geom xyz file with unit au
+        xyz_file = input_model.molecule.to_string(dtype='xyz', units='Angstrom')
+
+        # Create input dictionary
+        if template is None:
+            input_dict = {}
+            if input_model.driver == 'energy':
+                input_dict = {'dft': {
+                                'structure': {
+                                  'file': 'geometry.xyz'
+                                  },
+                                'xc': input_model.model.method,
+                                'ao': input_model.model.basis.upper(),
+                                'df_basis': input_model.keywords["df_basis"].upper(),
+                                'charge': input_model.molecule.molecular_charge
+                                },
+                              'print': {
+                                'results': True
+                                }
+                              }
+
+            # Write gradient call if asked for
+            if input_model.driver == 'gradient':
```

Comment:
`elif ..` here with an `else` exception at the end.

It looks like very little changes between inputs as well. Possible to reduce duplication?

Reply:
Yeah the structure between the two are slightly different, but I was able to reduce code duplication.